src/ape-together/ApeRebalanceExtension.sol

```diff
@@ -0,0 +1,265 @@
+import { PreciseUnitMath } from "indexcoop/contracts/lib/PreciseUnitMath.sol";
+
+import { SafeCast } from "@openzeppelin/contracts/utils/SafeCast.sol";
+import { Math } from "@openzeppelin/contracts/math/Math.sol";
+
+import { OwlNFT } from "./OwlNFT.sol";
+
+/**
+ * @title ApeRebalanceExtension
+ * @author Index Coop
+ *
+ * Rebalance extension that allows NFT holders to vote on token allocations. Rebalancing is still not
+ * fully trustless, as the operator must act as a price oracle for the tokens involved in the rebalance
+ * and set the trade data. However, a malicious operator can not introduce new tokens into the rebalance,
+ * but can influence the weighting by modifying the inputted price.
+ */
+contract ApeRebalanceExtension is GIMExtension {
+    using PreciseUnitMath for uint256;
+    using SafeCast for uint256;
+
+    /* ========== State Variables ======== */
+
+    uint256 public epochLength;
+    uint256 public  currentEpochStart;
+    
+    mapping(address => uint256) public lastEpochVoted;
+    mapping(address => uint256) public votes;
+    address[] public possibleComponents;
+
+    uint256 public immutable maxComponents;
+
+    OwlNFT public owlNft;
+
+    /* ========== Constructor ========== */
+
+    /**
+     * Sets state variables.
+     *
+     * @param _manager          address of manager contract
+     * @param _gim              address of Set Protocol GeneralIndexModule
+     * @param _owlNft           address of OwlNFT contract
+     * @param _startTime        timestamp for the start of the first voting period
+     * @param _epochLength      length of a voting period (in seconds)
+     * @param _maxComponents    maximum number of components in the set
+     */
+    constructor(
+        IBaseManager _manager,
+        IGeneralIndexModule _gim,
+        OwlNFT _owlNft,
+        uint256 _startTime,
+        uint256 _epochLength,
+        uint256 _maxComponents
+    )
+        public 
+        GIMExtension(_manager, _gim)
+    {
+        owlNft = _owlNft;
+        currentEpochStart = _startTime;
+        epochLength = _epochLength;
+        maxComponents = _maxComponents;
+    }
+
+    /* ======== External Functions ======== */
+ 
+    /**
+     * Submits a vote for an OwlNFT holder. Sum of all votes must not exceed the total
+     * votes that the NFT holder has alloted to them. This value can be fetched by calling
+     * getVotes on the NFT contract.
+     *
+     * @param _components   array of components that the NFT holder wants added to the index
+     * @param _votes        array of number of votes allocated to each of the components
+     */
+    function vote(address[] memory _components, uint256[] memory _votes) external {
+        require(_components.length == _votes.length, "length mismatch");
+        require(!_isEpochOver(), "voting period ended");
+
+        require(lastEpochVoted[msg.sender] != currentEpochStart, "already voted");
```

Comment:
Not seeing a check against the same NFT ID already being used to vote in the epoch. 

Is the vulnerability that an OwlNFT holder can transfer their NFT to another address they control and vote again? And do this repeatedly for as many addresses as they want.

Reply:
Yep!